core/offchain/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Substrate service. Starts a thread that spins up the network, client, and extrinsic pool.
+//! Manages communication between them.
+
+#![warn(missing_docs)]
+
+use std::{
+	marker::PhantomData,
+	sync::Arc,
+};
+
+use client::runtime_api::ApiExt;
+use inherents::pool::InherentsPool;
+use log::{debug, warn};
+use primitives::ExecutionContext;
+use runtime_primitives::{
+	generic::BlockId,
+	traits::{self, ProvideRuntimeApi},
+};
+use tokio::runtime::TaskExecutor;
+use transaction_pool::txpool::{Pool, ChainApi};
+
+mod api;
+
+pub use offchain_primitives::OffchainWorkerApi;
+
+/// An offchain workers manager.
+#[derive(Debug)]
+pub struct OffchainWorkers<C, Block: traits::Block> {
+	client: Arc<C>,
+	inherents_pool: Arc<InherentsPool<<Block as traits::Block>::Extrinsic>>,
+	executor: TaskExecutor,
+	_block: PhantomData<Block>,
+}
+
+impl<C, Block: traits::Block> OffchainWorkers<C, Block> {
+	/// Creates new `OffchainWorkers`.
+	pub fn new(
+		client: Arc<C>,
+		inherents_pool: Arc<InherentsPool<<Block as traits::Block>::Extrinsic>>,
+		executor: TaskExecutor,
+	) -> Self {
+		Self {
+			client,
+			inherents_pool,
+			executor,
+			_block: PhantomData,
+		}
+	}
+}
+
+impl<C, Block> OffchainWorkers<C, Block> where
+	Block: traits::Block,
+	C: ProvideRuntimeApi,
+	C::Api: OffchainWorkerApi<Block>,
+{
+	/// Start the offchain workers after given block.
+	pub fn on_block_imported<A>(
+		&self,
+		number: &<Block::Header as traits::Header>::Number,
+		pool: &Arc<Pool<A>>,
+	) where
+		A: ChainApi<Block=Block> + 'static,
+	{
+		let runtime = self.client.runtime_api();
+		let at = BlockId::number(*number);
+		let has_api = runtime.has_api::<OffchainWorkerApi<Block>>(&at);
+		debug!("Checking offchain workers at {:?}: {:?}", at, has_api);
+
+		if has_api.unwrap_or(false) {
+			let (api, runner) = api::Api::new(pool.clone(), self.inherents_pool.clone(), at.clone());
+			self.executor.spawn(runner.process());
```

Comment:
Sorry I don't know much about Rust futures, but can you ELI5 how this terminates?
Is the receiver returns `Ready` when tx is dropped?

Reply:
Yeah, so when the work runtime is doing is finished the `api` transmitting end will be dropped, that in turn will cause the `receiving.for_each` stream to be terminated, so the entire future returned from  `process()` will terminate.

But we currently allow the offchain workers to run arbitrarily long, it might make sense actually to kill them after some timeout, but that's something I don't really know how to do:
1. I think it's not possible to terminate wasm task currently
2. We could error when the worker attempts to use `ext_submit_extrinsic` (or even any other externality), but obviously it can also never happen.